### PR TITLE
[TT-11389]: moved graphql span attributes

### PR DESCRIPTION
**Pull request by kofoworola**
## **User description**
moved the document info attributes for graphql requests to the GraphqlMiddlewareValidation span

## Description
Moved the graphql-specific span attributes higher up the span chain so they can still be seen in the event of a request failure when detailed tracing is turned on

[TT-11389](https://tyktech.atlassian.net/browse/TT-11389)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

[TT-11389]: https://tyktech.atlassian.net/browse/TT-11389?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

___

## **Type**
enhancement

___

## **Description**
- Added OpenTelemetry semantic conventions for GraphQL operation name, type, and document in `ProcessAndStoreGraphQLRequest` method to enhance tracing capabilities.
- Refactored the way GraphQL operation type is determined and set on spans across different parts of the codebase.
- Removed redundant span attributes setting in the `ValidateForSchema` method of the detailed GraphQL engine.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>engine_v2.go</strong><dd><code>Add OpenTelemetry Attributes for GraphQL Requests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/graphengine/engine_v2.go
<li>Added OpenTelemetry semantic conventions for GraphQL in <br><code>ProcessAndStoreGraphQLRequest</code> method.<br> <li> Set attributes for GraphQL operation name, type, and document on the <br>span.

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6088/files#diff-b1eaa954c9836f395e1d49090e85c739e3878747c8bd748f556fc5a53ff7b191">+13/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>otel_graphql_engine_basic.go</strong><dd><code>Use PrintOperationType for Span Attributes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/graphql/otel_graphql_engine_basic.go
<li>Modified to use <code>PrintOperationType</code> for setting GraphQL operation type <br>attribute on the span.

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6088/files#diff-d2e2ab0951855338accb48c6985c5f880a8ea2746872c7cea5b64f308a5214c8">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>otel_graphql_engine_detailed.go</strong><dd><code>Refactor Span Attributes Setting in GraphQL Engine</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/graphql/otel_graphql_engine_detailed.go
<li>Removed span attributes setting from <code>ValidateForSchema</code> method.<br> <li> Exposed <code>PrintOperationType</code> function for external use.

</details>
    

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6088/files#diff-799c8fc0e5c351c4903c0ecfcc1b9997473f469edcd25118edf447dc6ea4e771">+2/-15</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by buger** [comment]
**API tests result - postgres15-sha256 env: success** :white_check_mark: 
Branch used: refs/pull/6088/merge
Commit:  
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/8137876446)

**Comment by buger** [comment]
**API tests result - mongo44-sha256 env: success** :white_check_mark: 
Branch used: refs/pull/6088/merge
Commit:  
Triggered by: pull_request (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/8137876446)

File: internal/graphengine/engine_v2.go [hunk 1]
```diff
@@ -214,6 +216,17 @@ func (e *EngineV2) ProcessAndStoreGraphQLRequest(w http.ResponseWriter, r *http.
 	if e.OpenTelemetry.Enabled && e.ApiDefinition.DetailedTracing {
 		ctx, span := e.OpenTelemetry.TracerProvider.Tracer().Start(r.Context(), "GraphqlMiddleware Validation")
 		defer span.End()
+
+		operationType, err := gqlRequest.OperationType()
+		if err != nil {
+			e.logger.Debug("error while getting operation type for trace", abstractlogger.Error(err))
```
Comment (buraksezer) [inline internal/graphengine/engine_v2.go:222]: It seems that we need to stop execution here by returning an error.

**Review by buraksezer — CHANGES_REQUESTED** [review]

File: internal/graphengine/engine_v2.go (hunk 1 above)
Comment (kofoworola) [reply internal/graphengine/engine_v2.go:222]: thanks, that was an oversight. Corrected

**Review by buraksezer — APPROVED** [review]

**Comment by buger** [comment]
**API tests result - mongo44-murmur64 env: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: 5ce6148b3591f71ffa8251cdf87849c4a41baa70 [TT-11389]: moved graphql span attributes (#6088)

## **User description**
moved the document info attributes for graphql requests to the
GraphqlMiddlewareValidation span



## Description
Moved the graphql-specific span attributes higher up the span chain so
they can still be seen in the event of a request failure when detailed
tracing is turned on

[TT-11389](https://tyktech.atlassian.net/browse/TT-11389)



## Related Issue






## Motivation and Context



## How This Has Been Tested






## Screenshots (if appropriate)

## Types of changes



- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test
coverage to functionality)

## Checklist





- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning
why it's required
- [ ] I would like a code coverage CI quality gate exception and have
explained why


[TT-11389]:
https://tyktech.atlassian.net/browse/TT-11389?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ


___

## **Type**
enhancement


___

## **Description**
- Added OpenTelemetry semantic conventions for GraphQL operation name,
type, and document in `ProcessAndStoreGraphQLRequest` method to enhance
tracing capabilities.
- Refactored the way GraphQL operation type is determined and set on
spans across different parts of the codebase.
- Removed redundant span attributes setting in the `ValidateForSchema`
method of the detailed GraphQL engine.


___



## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant
files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
<summary><strong>engine_v2.go</strong><dd><code>Add OpenTelemetry
Attributes for GraphQL Requests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;
&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/graphengine/engine_v2.go
<li>Added OpenTelemetry semantic conventions for GraphQL in
<br><code>ProcessAndStoreGraphQLRequest</code> method.<br> <li> Set
attributes for GraphQL operation name, type, and document on the
<br>span.


</details>
    

  </td>
<td><a
href="https://github.com/TykTechnologies/tyk/pull/6088/files#diff-b1eaa954c9836f395e1d49090e85c739e3878747c8bd748f556fc5a53ff7b191">+13/-0</a>&nbsp;
&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
<summary><strong>otel_graphql_engine_basic.go</strong><dd><code>Use
PrintOperationType for Span Attributes</code>&nbsp; &nbsp; &nbsp; &nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;
&nbsp; </dd></summary>
<hr>

internal/graphql/otel_graphql_engine_basic.go
<li>Modified to use <code>PrintOperationType</code> for setting GraphQL
operation type <br>attribute on the span.


</details>
    

  </td>
<td><a
href="https://github.com/TykTechnologies/tyk/pull/6088/files#diff-d2e2ab0951855338accb48c6985c5f880a8ea2746872c7cea5b64f308a5214c8">+1/-1</a>&nbsp;
&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>

<summary><strong>otel_graphql_engine_detailed.go</strong><dd><code>Refactor
Span Attributes Setting in GraphQL Engine</code>&nbsp; &nbsp; &nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/graphql/otel_graphql_engine_detailed.go
<li>Removed span attributes setting from <code>ValidateForSchema</code>
method.<br> <li> Exposed <code>PrintOperationType</code> function for
external use.


</details>
    

  </td>
<td><a
href="https://github.com/TykTechnologies/tyk/pull/6088/files#diff-799c8fc0e5c351c4903c0ecfcc1b9997473f469edcd25118edf447dc6ea4e771">+2/-15</a>&nbsp;
&nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools
and their descriptions
Triggered by: push (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/8190304325)

**Comment by buger** [comment]
**API tests result - postgres15-murmur64 env: success** :white_check_mark: 
Branch used: refs/heads/master
Commit: 5ce6148b3591f71ffa8251cdf87849c4a41baa70 [TT-11389]: moved graphql span attributes (#6088)

## **User description**
moved the document info attributes for graphql requests to the
GraphqlMiddlewareValidation span



## Description
Moved the graphql-specific span attributes higher up the span chain so
they can still be seen in the event of a request failure when detailed
tracing is turned on

[TT-11389](https://tyktech.atlassian.net/browse/TT-11389)



## Related Issue






## Motivation and Context



## How This Has Been Tested






## Screenshots (if appropriate)

## Types of changes



- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing
functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test
coverage to functionality)

## Checklist





- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning
why it's required
- [ ] I would like a code coverage CI quality gate exception and have
explained why


[TT-11389]:
https://tyktech.atlassian.net/browse/TT-11389?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ


___

## **Type**
enhancement


___

## **Description**
- Added OpenTelemetry semantic conventions for GraphQL operation name,
type, and document in `ProcessAndStoreGraphQLRequest` method to enhance
tracing capabilities.
- Refactored the way GraphQL operation type is determined and set on
spans across different parts of the codebase.
- Removed redundant span attributes setting in the `ValidateForSchema`
method of the detailed GraphQL engine.


___



## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant
files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
<summary><strong>engine_v2.go</strong><dd><code>Add OpenTelemetry
Attributes for GraphQL Requests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;
&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/graphengine/engine_v2.go
<li>Added OpenTelemetry semantic conventions for GraphQL in
<br><code>ProcessAndStoreGraphQLRequest</code> method.<br> <li> Set
attributes for GraphQL operation name, type, and document on the
<br>span.


</details>
    

  </td>
<td><a
href="https://github.com/TykTechnologies/tyk/pull/6088/files#diff-b1eaa954c9836f395e1d49090e85c739e3878747c8bd748f556fc5a53ff7b191">+13/-0</a>&nbsp;
&nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
<summary><strong>otel_graphql_engine_basic.go</strong><dd><code>Use
PrintOperationType for Span Attributes</code>&nbsp; &nbsp; &nbsp; &nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;
&nbsp; </dd></summary>
<hr>

internal/graphql/otel_graphql_engine_basic.go
<li>Modified to use <code>PrintOperationType</code> for setting GraphQL
operation type <br>attribute on the span.


</details>
    

  </td>
<td><a
href="https://github.com/TykTechnologies/tyk/pull/6088/files#diff-d2e2ab0951855338accb48c6985c5f880a8ea2746872c7cea5b64f308a5214c8">+1/-1</a>&nbsp;
&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>

<summary><strong>otel_graphql_engine_detailed.go</strong><dd><code>Refactor
Span Attributes Setting in GraphQL Engine</code>&nbsp; &nbsp; &nbsp;
&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

internal/graphql/otel_graphql_engine_detailed.go
<li>Removed span attributes setting from <code>ValidateForSchema</code>
method.<br> <li> Exposed <code>PrintOperationType</code> function for
external use.


</details>
    

  </td>
<td><a
href="https://github.com/TykTechnologies/tyk/pull/6088/files#diff-799c8fc0e5c351c4903c0ecfcc1b9997473f469edcd25118edf447dc6ea4e771">+2/-15</a>&nbsp;
&nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> ✨ **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools
and their descriptions
Triggered by: push (@kofoworola)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/8190304325)

**Pull request merged**